### PR TITLE
Static Analysis With Psalm

### PR DESCRIPTION
**Pull request by tarecord**
## What does this do/fix?
Adds [Psalm](https://github.com/vimeo/psalm/) to handle static analysis of the codebase. Right now, it's only consuming and reviewing `wp-content/plugins/core/src`. It's also pulling in a [Human Made plugin](https://github.com/humanmade/psalm-plugin-wordpress) that configures Psalm to use WordPress stubs so it doesn't throw errors for missing functions/types.

## Todo
- [ ] ~Clean up current errors with the core plugin to start with a "clean slate" with new development and avoid failing builds~ Psalm provides a way to set up a baseline file to ignore current errors—if you don't have time to fix them all.
- [x] Set up Github Actions to automatically run Psalm checks

### CODE REVIEW

**Comment by nickpelton** [comment]
Neat! 

I think we should also add the path for the theme and cover all our generated code. I would also make sure we address all the PHP errors we'll see before adding the GIthub action otherwise we'll be getting failed builds. Are you planning on addressing the code issues in this branch? Or another one?

**Comment by tarecord** [comment]
I totally agree, I wanted to get a PR up to start the discussion.

I was thinking about handling the current errors in another branch for now, and merging that into this branch so we can keep the commit history clean.

**Comment by nickpelton** [comment]
Sounds like a good appoach!

**Comment by defunctl** [comment]
Love this! We should only commit a `psalm.xml.dist` and put the `pslam.xml` in .gitignore to allow developers to customize/change it locally if needed.

Would this be something that could be integrated into husky hooks library we have to automatically run?

**Comment by tarecord** [comment]
Psalm provides a way to ignore all the current errors in the codebase so you don't have to spend a massive amount of time fixing all the issues upfront. They can be fixed moving forward and we can begin using Psalm with new code:
>If you have a bunch of errors and you don't want to fix them all at once, Psalm can grandfather-in errors in existing code, while ensuring that new code doesn't have those same sorts of errors.
```
vendor/bin/psalm --set-baseline=your-baseline.xml
```
>will generate a file containing the current errors. You can commit that generated file so that Psalm running in other places (e.g. CI) won't complain about those errors either, and you can update that baseline file (to remove references to things that have been fixed) with
```
vendor/bin/psalm --update-baseline
```
>Baseline files are a great way to gradually improve a codebase.

File: wp-content/plugins/core/src/Admin/Admin_Subscriber.php
Comment (defunctl) [inline wp-content/plugins/core/src/Admin/Admin_Subscriber.php:40]: Is there any way to make it ignore the args in the closure, these should not be typed because WordPress does not send reliable types.

It should be like this:

```php
add_filter( 'teeny_mce_buttons', function ( $buttons, $editor_id ): array {
			return $this->container->get(Classic_Editor_Formats::class)->teeny_mce_buttons( (array) $buttons, (string) $editor_id );
		}, 10, 2 );
```

File: wp-content/plugins/core/src/Blocks/Types/Buttons/Buttons.php
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Buttons/Buttons.php:27]: Weird it didn't add `: void` here and below

Comment (tarecord) [reply wp-content/plugins/core/src/Blocks/Types/Buttons/Buttons.php:27]: It looks at PHPDoc tags to detect types, but I wonder if there's a way to force it to add the `: type` syntax. https://psalm.dev/docs/annotating_code/supported_annotations/

Comment (tarecord) [reply wp-content/plugins/core/src/Blocks/Types/Buttons/Buttons.php:27]: Seems like we only need this to force the syntax.
```xml
<psalm
  useDocblockTypes="false"
>
```

File: wp-content/plugins/core/src/Admin/Admin_Subscriber.php
Comment (tarecord) [reply wp-content/plugins/core/src/Admin/Admin_Subscriber.php:40]: we _could_ [globally ignore closure return types](https://psalm.dev/docs/running_psalm/issues/MissingClosureReturnType/) with `MissingClosureReturnType` in the [config suppression](https://psalm.dev/docs/running_psalm/dealing_with_code_issues/#config-suppression), but I'm sure we use them elsewhere we'd like to check types.

Comment (defunctl) [reply wp-content/plugins/core/src/Admin/Admin_Subscriber.php:40]: Ahh that rule is for the return type of the closure, not the parameters being passed into it.

File: wp-content/plugins/core/src/Blocks/Types/Buttons/Buttons.php
Comment (defunctl) [reply wp-content/plugins/core/src/Blocks/Types/Buttons/Buttons.php:27]: Nice, the new PHPCS rules enforce this as well, so might not be needed?

**Comment by nickpelton** [comment]
Giving this a test today and have a few questions. PHPCS also does a lot of these checks. How do we avoid doubling up and maintaining two possibly conflicting rule sets?

**Case**: Currently after  merging main, the commit hook PHPcs check won't allow commits after conflicting Psalm changes.

```Found existing alias for "git commit -m". You should use: "gcmsg"
⠴ waiting ^R                                                                                                                                                                                           
  EXECUTE > phpcs-core-plugin
 ...............................E...E.............E.......... 60 / 65 (92%)
.....                                                        65 / 65 (100%)



FILE: wp-content/plugins/core/src/Admin/Admin_Subscriber.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 1 LINE
----------------------------------------------------------------------
 42 | ERROR | [x] Expected 1 spaces after opening parenthesis; 0
    |       |     found
    |       |     (PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket)
 42 | ERROR | [x] Expected 1 spaces before closing parenthesis; 0
    |       |     found
    |       |     (PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...plugins/core/src/Integrations/Google_Tag_Manager/GTM_Scripts.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 57 | ERROR | [x] Expected 1 line between description and
    |       |     annotations, found 0.
    |       |     (SlevomatCodingStandard.Commenting.DocCommentSpacing.IncorrectLinesCountBetweenDescriptionAndAnnotations)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: wp-content/plugins/core/src/Shortcodes/Gallery.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 108 | ERROR | [x] Referencing general \Exception; use \Throwable
     |       |     instead.
     |       |     (SlevomatCodingStandard.Exceptions.ReferenceThrowableOnly.ReferencedGeneralException)
 121 | ERROR | [x] There must be one empty line before class closing
     |       |     brace.
     |       |     (SlevomatCodingStandard.Classes.EmptyLinesAroundClassBraces.NoEmptyLineBeforeClosingBrace)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------

Time: 3.64 secs; Memory: 12MB



SUMMARY: (done in 3.93 seconds)
✔️  phpcs-mu
✔️  phpcs-core-theme
🥊  phpcs-core-plugin
```

**Review by nickpelton — CHANGES_REQUESTED** [review]
Incompatible with existing PHPcs rules.

**Comment by nickpelton** [comment]
Another note. The GitHub actions Workflow name matches the existing `PHP Composer and Codesniffer` action. We should rename it if this is a separate analysis task. OR we should put them together since both require a build step and we can save CI minutes.

File: .github/workflows/static-analysis.yml
Comment (defunctl) [inline .github/workflows/static-analysis.yml:27]:
```suggestion
          TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

File: composer.json
Comment (defunctl) [inline composer.json:136]: What's this doing in here?

File: wp-content/plugins/core/src/Blocks/Types/Card_Grid/Card_Grid.php
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Card_Grid/Card_Grid.php:32]: Not sure what's up with spacing on these comments or if I did that in another PR by accident?

File: wp-content/plugins/core/src/Blocks/Types/Hero/Hero.php
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Hero/Hero.php:28]: Yeah, not sure what's going on here?

File: wp-content/plugins/core/src/Blocks/Types/Interstitial/Interstitial.php
Comment (defunctl) [inline wp-content/plugins/core/src/Blocks/Types/Interstitial/Interstitial.php:28]: Would be way better to just add `add_block(): void {` to all of these remove the @return docBlock. We don't need return comments if the method specifies what they are unless it's an array and then we should specify what type of array it is.

File: wp-content/plugins/core/src/Cache/Listener.php
Comment (defunctl) [inline wp-content/plugins/core/src/Cache/Listener.php:34]:
```suggestion
	public function p2p_created_connection( $p2p_id ): void {
```

File: wp-content/plugins/core/src/Integrations/Google_Tag_Manager/GTM_Scripts.php
Comment (defunctl) [inline wp-content/plugins/core/src/Integrations/Google_Tag_Manager/GTM_Scripts.php:19]: Is it psalm that's mangling all of these docblocks?

**Review by defunctl — CHANGES_REQUESTED** [review]
Can we configure this to only specify @return docblocks when it's an array or we are not returning a specific type defined method type hints?

**Review by dpellenwood — COMMENTED** [review]
This is a bit out of my wheel house, so I'm going to defer to others on this PR.

File: composer.json
Comment (tarecord) [reply composer.json:134]: not sure what you mean by this?

**Comment by nickpelton** [comment]
Hey @tarecord , I still this this would be a valuable addition. Do you think you'll have time to work on this in the near future?

**Comment by tarecord** [comment]
Closing in favor of #971

**Pull request closed without merging**
